src/librustdoc/html/render/context.rs

```diff
@@ -409,24 +415,36 @@ impl<'tcx> FormatRenderer<'tcx> for Context<'tcx> {
             current: Vec::new(),
             dst,
             render_redirect_pages: false,
-            id_map: Rc::new(RefCell::new(id_map)),
-            deref_id_map: Rc::new(RefCell::new(FxHashMap::default())),
-            shared: Arc::new(scx),
-            all: Rc::new(RefCell::new(AllTypes::new())),
-            errors: Rc::new(receiver),
+            id_map: RefCell::new(id_map),
+            deref_id_map: RefCell::new(FxHashMap::default()),
+            shared: Rc::new(scx),
             cache: Rc::new(cache),
-            redirections: if generate_redirect_map { Some(Default::default()) } else { None },
         };
 
         CURRENT_DEPTH.with(|s| s.set(0));
 
         // Write shared runs within a flock; disable thread dispatching of IO temporarily.
-        Arc::get_mut(&mut cx.shared).unwrap().fs.set_sync_only(true);
+        Rc::get_mut(&mut cx.shared).unwrap().fs.set_sync_only(true);
         write_shared(&cx, &krate, index, &md_opts)?;
-        Arc::get_mut(&mut cx.shared).unwrap().fs.set_sync_only(false);
+        Rc::get_mut(&mut cx.shared).unwrap().fs.set_sync_only(false);
         Ok((cx, krate))
     }
 
+    fn make_child_renderer(&self) -> Self {
+        let mut id_map = IdMap::new();
+        id_map.populate(&INITIAL_IDS);
```

Comment:
I wonder about this change: before we were using `reset()` on the `IdMap`, which, I assume, doesn't change its memory size, just removes the indexes. This is a change of behavior, which might be the reason there is a little performance loss, no?

Reply:
`reset` is defined as

```rust
crate fn reset(&mut self) {
    self.map = init_id_map();
}
```

which appears to *not* reuse the allocation; instead it allocates a fresh hash map. I then checked [the Assembly](https://rust.godbolt.org/z/P6baMa) of

```rust
pub fn foo(hash: &mut HashMap<String, usize>) {
    *hash = HashMap::new();
}
```

and I see a call to `__rust_dealloc`, which seems to confirm that the allocation is not reused in `IdMap::reset()`.

Reply:
Perfect then! That concludes the last thing I was unclear on.